src/main/java/appeng/tile/misc/TileCharger.java

```diff
@@ -215,55 +215,79 @@ public TickRateModulation tickingRequest( IGridNode node, int TicksSinceLastCall
 	private boolean doWork()
 	{
 		final ItemStack myItem = this.inv.getStackInSlot( 0 );
+		boolean changed = false;
 
-		// charge from the network!
-		if( this.getInternalCurrentPower() < 1499 )
+		if( !myItem.isEmpty() )
 		{
-			try
-			{
-				this.injectExternalPower( PowerUnits.AE, this.getProxy().getEnergy().extractAEPower( Math.min( 500.0, 1500.0 - this.getInternalCurrentPower() ),
-						Actionable.MODULATE, PowerMultiplier.ONE ), Actionable.MODULATE );
-			}
-			catch( final GridAccessException e )
+			final IMaterials materials = AEApi.instance().definitions().materials();
+
+			if( Platform.isChargeable( myItem ) )
 			{
-				// continue!
-			}
+				final IAEItemPowerStorage ps = (IAEItemPowerStorage) myItem.getItem();
 
-			return true;
-		}
+				if( ps.getAEMaxPower( myItem ) > ps.getAECurrentPower( myItem ) )
+				{
+					final double chargeRate = AEApi.instance().registries().charger().getChargeRate( myItem.getItem() );
 
-		if( myItem.isEmpty() )
-		{
-			return false;
-		}
+					double extractedAmount = this.extractAEPower( chargeRate, Actionable.MODULATE, PowerMultiplier.CONFIG );
 
-		final IMaterials materials = AEApi.instance().definitions().materials();
+					final double missingChargeRate = chargeRate - extractedAmount;
+					final double missingAEPower = ps.getAEMaxPower( myItem ) - ps.getAECurrentPower( myItem );
+					final double toExtract = Math.min( missingChargeRate, missingAEPower );
 
-		if( this.getInternalCurrentPower() > 149 && Platform.isChargeable( myItem ) )
-		{
-			final IAEItemPowerStorage ps = (IAEItemPowerStorage) myItem.getItem();
-			if( ps.getAEMaxPower( myItem ) > ps.getAECurrentPower( myItem ) )
-			{
-				final double oldPower = this.getInternalCurrentPower();
+					try
+					{
+						extractedAmount += this.getProxy().getEnergy().extractAEPower( toExtract, Actionable.MODULATE, PowerMultiplier.ONE );
+					}
+					catch( GridAccessException e1 )
+					{
+						// Ignore.
+					}
 
-				final double adjustment = ps.injectAEPower( myItem, this.extractAEPower( 150.0, Actionable.MODULATE, PowerMultiplier.CONFIG ),
-						Actionable.MODULATE );
-				this.setInternalCurrentPower( this.getInternalCurrentPower() + adjustment );
+					if( extractedAmount > 0 )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)